CLOUDP-347194 - enable Pod Security Admission at `restricted` level

MaciejKaras opened the pull request:
# Summary

Based on the [HELP-81729](https://jira.mongodb.org/browse/HELP-81729) ticket I investigated if our workloads align with `restricted` [Pod Security Standards](https://v1-32.docs.kubernetes.io/docs/concepts/security/pod-security-standards/) security level. Enforcing `restricted` cannot be guaranteed across our e2e setup — for example Istio injects `istio-init` containers that do not follow the restricted profile, and our `tests` pod does not meet the requirements either.

Because of this the `warn` mode for the `restricted` level is enabled instead of `enforce`. For the complex `e2e_om_ops_manager_backup_sharded_cluster` test, `restricted` is enforced on a single (non-multi) cluster so we can monitor PSS alignment. More on levels and modes [here](https://v1-32.docs.kubernetes.io/docs/concepts/security/pod-security-admission/).

Concretely, the more secure default `securityContext` settings applied are:
- Operator deployment: `allowPrivilegeEscalation: false`, `capabilities.drop: [ALL]`, `seccompProfile.type: RuntimeDefault`
- Other workloads: `capabilities.drop: [ALL]` (container level), `seccompProfile.type: RuntimeDefault` (pod level)

If less restrictive settings are required, use `template` / `podTemplate` overrides.

- **Note on the latest revision**: commit `2d5ef4e` ("Apply remaining changes") inadvertently reset all 14 substantive files back to `master`, discarding the hardening and leaving only the changelog. This revision restores the full PSS work from parent `125366e`. The original request that triggered it — fixing lint issues — required no changes: `black`, `isort`, `shellcheck`, `gofmt`, and `go build` all pass clean on the PR-touched files (the remaining repo-wide `golangci-lint` and govulncheck findings are pre-existing in unmodified files, the latter tracked in #1189).

Three Security issues can be closed this PR:

- https://jira.mongodb.org/browse/ENGSEC-822
- https://jira.mongodb.org/browse/ENGSEC-821
- https://jira.mongodb.org/browse/ENGSEC-820

## Proof of Work

Passing CI for all tests that have warning + passing enforcement for `e2e_om_ops_manager_backup_sharded_cluster` test.

Example [warning](https://parsley-beta.corp.mongodb.com/evergreen/mongodb_kubernetes_e2e_multi_cluster_om_appdb_e2e_om_ops_manager_backup_tls_patch_58f35690937269c5e6028d08a3abe3f49d0d09a1_68d3e3c4a81aa80007f67b5f_25_09_24_12_27_51/0/task?bookmarks=0,34024&shareLine=33307) from the `e2e_om_ops_manager_backup_tls` test:
```
[2025/09/24 14:59:58.691] Warning: would violate PodSecurity "restricted:latest": unrestricted capabilities (container "istio-init" must not include "NET_ADMIN", "NET_RAW" in securityContext.capabilities.add), runAsNonRoot != true (container "istio-init" must not set securityContext.runAsNonRoot=false), runAsUser=0 (container "istio-init" must not set runAsUser=0)
```

## Checklist

- [x] Have you linked a jira ticket and/or is the ticket in the title?
- [x] Have you checked whether your jira ticket required DOCSP changes?
- [x] Have you added changelog file?
    - use `skip-changelog` label if not needed
    - refer to [Changelog files and Release Notes](https://github.com/mongodb/mongodb-kubernetes/blob/master/CONTRIBUTING.md#changelog-files-and-release-notes) section in CONTRIBUTING.md for more details

MaciejKaras committed:
CLOUDP-347194 - enable Pod Security Admission at `restricted` level

MaciejKaras committed:
Fix unit tests + linter issues

MaciejKaras committed:
Disable enforce for multi-cluster test

MaciejKaras committed:
Add changelog file

MaciejKaras committed:
Fix enforcing the restricted mode

MaciejKaras committed:
Update the changelog

lsierant commented on controllers/operator/construct/database_construction_test.go:
```diff
@@ -45,6 +45,9 @@ func Test_buildDatabaseInitContainer(t *testing.T) {
 		SecurityContext: &corev1.SecurityContext{
 			ReadOnlyRootFilesystem:   ptr.To(true),
 			AllowPrivilegeEscalation: ptr.To(false),
+			Capabilities: &corev1.Capabilities{
```
is there any potential that adding this default will break *any* customer's workload and will require some manual intervention? Just thinking about our semver guarantees.

mircea-cosbuc replied:
This is our defaults. Only problem I see with this is some customers now requiring explicitly setting capabilities.

lsierant replied:
Still, do we force customers to do any manual fix when upgrading? If yes, we need to bump major.

MaciejKaras replied:
I don't understand what will we break here. We are changing our default SecurityContext for operator and other pods created. If customer wants to have dedicated SecurityContext or PodSecurityContext they need to set `MANAGED_SECURITY_CONTEXT` env var and our defaults will be completely overwritten. If they don't set `MANAGED_SECURITY_CONTEXT` every change they make to SecurityContext manually will be overwritten by our defaults.

MaciejKaras committed:
Add more details about overrides in the changelog

MaciejKaras committed:
Update changelog syntax

vinilage reviewed (approved):
LGTM!

MaciejKaras committed:
Merge branch 'master' into maciejk/pss-warn

# Conflicts:
#	helm_chart/templates/operator.yaml
#	scripts/evergreen/deployments/test-app/templates/mongodb-enterprise-tests.yaml

MaciejKaras committed:
Merge branch 'master' into maciejk/pss-warn

MaciejKaras committed:
Merge branch 'master' into maciejk/pss-warn

# Conflicts:
#	docker/mongodb-kubernetes-tests/tests/opsmanager/om_ops_manager_backup_sharded_cluster.py
#	helm_chart/tests/operator_security_context_test.yaml
#	scripts/evergreen/deployments/test-app/templates/mongodb-enterprise-tests.yaml

MaciejKaras committed:
pre-commit fixes

MaciejKaras committed:
Merge branch 'master' into maciejk/pss-warn

# Conflicts:
#	scripts/dev/prepare-multi-cluster/main.go

MaciejKaras committed:
Update changelog layout

MaciejKaras committed:
Merge branch 'master' into maciejk/pss-warn

MaciejKaras committed:
fix: remove redundant update_standalone causing pre-commit infinite loop

update_standalone_installer patched public/*.yaml with the operator image
version, but generate_standalone_yaml (the next pre-commit hook) regenerates
those same files via `helm template -f helm_chart/values.yaml`, which already
embeds the correct version from values.yaml. The two hooks fought on every run:
update_standalone_installer reformatted sequence indentation with ruamel.yaml
offset=2, then generate_standalone_yaml overwrote with helm's native format.

Remove update_standalone and its now-dead update_standalone_installer helper.

MaciejKaras committed:
Merge branch 'fix/pre-commit-update-standalone-loop' into maciejk/pss-warn

MaciejKaras committed:
fix: regenerate standalone YAMLs with corrected capabilities.drop indentation

lucian-tosa commented on docker/mongodb-kubernetes-tests/tests/opsmanager/om_ops_manager_backup_sharded_cluster.py:
```diff
@@ -33,15 +33,25 @@ def s3_bucket(aws_s3_client: AwsS3Client, namespace: str) -> Iterator[str]:
 
+@fixture(scope="module")
+def enforced_pss_namespace(namespace: str) -> str:
+    # Change pod-security mode from warn to enforce. This will make test fail if operator and deployments don't support enforce mode
+    # This will not work in multi-cluster, because Istio injects sidecar and that breaks restricted level
```
as a sanity check, we can enforce the restricted pss on a no_mesh multi-cluster test

MaciejKaras replied:
I was able to actually enforce the pss restricted for mesh mc tests as well in the https://github.com/mongodb/mongodb-kubernetes/pull/1199/changes. The plan is to merge it afterwards.

lucian-tosa commented on helm_chart/tests/operator_security_context_test.yaml:
```diff
@@ -87,8 +99,22 @@ tests:
+          path: spec.template.spec.containers[?(@.name=="mongodb-kubernetes-operator-multi-cluster")].securityContext.capabilities.drop
+      - equal:
+          path: spec.template.spec.containers[?(@.name=="mongodb-kubernetes-operator-multi-cluster")].securityContext.capabilities.drop
+          value: [ALL]
```
I'm very confused as to how this changed (and it passes) if values-multi-cluster.yaml and the template remains unchanged

MaciejKaras replied:
That's weird, I'll investigate it further

MaciejKaras replied:
It's really weird, the operator.yaml did change, but the PR diff in GitHub UI does not show it !

MaciejKaras replied:
Ok, I see it now. The changes to security-context are present only in the values.yaml. If you pass an extra file like -f values-specific.yaml, Helm merges it on top of the default values. So values-specific.yaml is an override layer, not a replacement.

I've tested it locally and verified it.

lucian-tosa commented on docker/mongodb-kubernetes-tests/kubetester/helm.py:
```diff
@@ -36,7 +36,8 @@ def helm_template(
 
-    yaml_file_name = "{}.yaml".format(str(uuid.uuid4()))
+    home = os.getenv("HOME")
+    yaml_file_name = os.path.join(home, "{}.yaml".format(str(uuid.uuid4())))
```
this would create conflicts if running on different worktrees

MaciejKaras replied:
Nice catch!

MaciejKaras replied:
Reverted completely

MaciejKaras committed:
revert helm.py changes

MaciejKaras committed:
fix: write helm template output to /tmp to avoid readOnlyRootFilesystem permission error

Replace uuid-based CWD write with tempfile.mkstemp so the file lands in /tmp,
which is a writable tmpfs even when PSS restricted enforces readOnlyRootFilesystem.
Also avoids $HOME pollution and worktree conflicts.

MaciejKaras committed:
Merge branch 'master' into maciejk/pss-warn
